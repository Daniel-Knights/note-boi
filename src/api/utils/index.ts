export * from './decorators';
export * from './error';
export * from './http';
