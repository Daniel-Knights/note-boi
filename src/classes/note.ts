import type Delta from 'quill-delta';

export class Note {
  readonly uuid: string = crypto.randomUUID();
  timestamp = Date.now();
  content: NoteContent = {
    title: '',
    body: '',
    delta: {
      ops: [],
    },
  };
}

type NoteContent = {
  title: string;
  body: string;
  delta: Partial<Delta>;
};
