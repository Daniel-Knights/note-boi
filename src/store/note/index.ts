export * from './event';
export * from './operations';
export * from './state';
export * from './types';
export * from './utils';
