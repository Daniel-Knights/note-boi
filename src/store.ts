import { reactive } from 'vue';
import { invoke } from '@tauri-apps/api/tauri';
import { v4 as uuidv4 } from 'uuid';

import { isEmptyNote } from './utils';

export class Note {
  readonly id = uuidv4();
  body = {
    delta: '',
    text: '',
  };
  timestamp = Date.now();
}

interface State {
  notes: Note[];
  selectedNote: Note;
}

export const state = reactive<State>({
  notes: [],
  selectedNote: new Note(),
});

/** Sorts notes in descending order by timestamp. */
function sortNotes() {
  state.notes.sort((a, b) => b.timestamp - a.timestamp);
}

/** Finds note index within {@link state.notes}. */
function findNoteIndex(id: string) {
  return state.notes.findIndex((nt) => nt.id === id);
}

/** Finds note within {@link state.notes}. */
export function findNote(id: string): Note | undefined {
  return state.notes.find((nt) => nt.id === id);
}

/** Deletes {@link state.selectedNote} when both `title` and `body` are empty. */
function clearEmptyNote(): void {
  const isValidClear = state.notes.length > 1;

  const foundNote = findNote(state.selectedNote.id);
  if (!foundNote) return;

  if (isValidClear && isEmptyNote(foundNote)) {
    deleteNote(state.selectedNote.id);
  }
}

const selectEvent = new CustomEvent('note-select');

/**
 * Looks for note with given `id` in {@link state.notes},
 * and sets it to {@link state.selectedNote}.
 */
export function selectNote(id: string): void {
  if (state.selectedNote.id === id) return;

  clearEmptyNote();

  const foundNote = findNote(id);
  if (foundNote) state.selectedNote = { ...foundNote };

  document.dispatchEvent(selectEvent);
}

/** Fetches all notes and updates {@link state}. */
export async function getAllNotes(): Promise<void> {
  const fetchedNotes = await invoke<Note[]>('get_all_notes').catch(console.error);

  const hasNotes = fetchedNotes && fetchedNotes.length > 0;
  if (!hasNotes) return newNote();

  state.notes = fetchedNotes;

  sortNotes();

  if (isEmptyNote(fetchedNotes[0])) {
    state.notes[0].timestamp = Date.now();
  }

  state.selectedNote = { ...state.notes[0] };

  document.dispatchEvent(selectEvent);
}

/** Deletes note with the given `id`. */
export function deleteNote(id: string): void {
  state.notes.splice(findNoteIndex(id), 1);
  state.selectedNote = { ...state.notes[0] };

  sortNotes();

  invoke('delete_note', { id }).catch(console.error);
}

/** Creates an empty note. */
export function newNote(): void {
  const foundNote = findNote(state.selectedNote.id);

  // Only update timestamp if selected note is empty
  if (foundNote && isEmptyNote(foundNote)) {
    state.selectedNote.timestamp = Date.now();
    return;
  }

  const freshNote = new Note();

  state.notes.unshift(freshNote);
  state.selectedNote = { ...freshNote };

  invoke('new_note', { note: { ...freshNote } }).catch(console.error);

  document.dispatchEvent(selectEvent);
}

/** Edits note body on Quill `text-change`. */
export function editBody(delta: string, text: string): void {
  const foundNote = findNote(state.selectedNote.id);
  if (!foundNote || text === foundNote.body.text) return;

  const timestamp = Date.now();
  foundNote.timestamp = timestamp;
  state.selectedNote.timestamp = timestamp;

  foundNote.body = { delta, text };

  sortNotes();

  invoke('edit_note', { note: { ...foundNote } }).catch(console.error);
}
