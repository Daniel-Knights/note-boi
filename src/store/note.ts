import * as dialog from '@tauri-apps/plugin-dialog';
import type Delta from 'quill-delta';
import { reactive } from 'vue';

import { NOTE_EVENTS } from '../constant';
import { isEmptyNote, tauriInvoke } from '../utils';

import { autoPush, syncState, UnsyncedNoteIds } from './sync';

export type UnsyncedEventDetail = {
  noteId: string;
  kind: keyof Pick<UnsyncedNoteIds, 'edited' | 'deleted'>;
};

export class Note {
  readonly id: string = crypto.randomUUID();
  timestamp = Date.now();
  content = {
    delta: <Partial<Delta>>{
      ops: [],
    },
    title: '',
    body: '',
  };
}

export const noteState = reactive({
  notes: <Note[]>[],
  selectedNote: new Note(),
  extraSelectedNotes: <Note[]>[],
});

const newNoteEvent = new CustomEvent(NOTE_EVENTS.new);
export const selectNoteEvent = new CustomEvent(NOTE_EVENTS.select);
export const changeNoteEvent = new CustomEvent(NOTE_EVENTS.change);

function getUnsyncedEvent(
  noteId: UnsyncedEventDetail['noteId'],
  kind: UnsyncedEventDetail['kind']
) {
  return new CustomEvent(NOTE_EVENTS.unsynced, { detail: { noteId, kind } });
}

/** Sorts notes in descending order by timestamp. */
export const sortNotesFn = (a: Note, b: Note): number => b.timestamp - a.timestamp;

/** Sorts {@link noteState.notes} in descending order by timestamp. */
export const sortStateNotes = (): Note[] => noteState.notes.sort(sortNotesFn);

/** Finds note index within {@link noteState.notes}. */
export function findNoteIndex(id?: string): number {
  return noteState.notes.findIndex((nt) => nt.id === id);
}

/** Finds note within {@link noteState.notes}. */
export function findNote(id?: string): Note | undefined {
  return noteState.notes.find((nt) => nt.id === id);
}

/** Deletes {@link noteState.selectedNote} when note is empty. */
function clearEmptyNote(): void {
  const isValidClear = noteState.notes.length > 1;
  if (!isValidClear) return;

  const foundNote = findNote(noteState.selectedNote.id);
  if (!foundNote) return;

  if (isEmptyNote(foundNote)) {
    deleteNote(noteState.selectedNote.id);
  }
}

/**
 * Looks for note with given `id` in {@link noteState.notes},
 * and sets it to {@link noteState.selectedNote}.
 */
export function selectNote(id?: string): void {
  if (noteState.selectedNote.id === id) return;

  clearEmptyNote();

  const foundNote = findNote(id);
  if (!foundNote) return;

  noteState.selectedNote = { ...foundNote };

  document.dispatchEvent(selectNoteEvent);
  document.dispatchEvent(changeNoteEvent);
}

/**
 * Returns true if note is either {@link noteState.selectedNote}
 * or within {@link noteState.extraSelectedNotes}.
 */
export function isSelectedNote(note: Note): boolean {
  return (
    note.id === noteState.selectedNote.id ||
    noteState.extraSelectedNotes.some((nt) => nt?.id === note.id)
  );
}

/** Fetches all notes and updates {@link noteState}. */
export async function getAllNotes(): Promise<void> {
  const fetchedNotes = await tauriInvoke('get_all_notes').catch(console.error);

  const hasNotes = fetchedNotes && fetchedNotes.length > 0;
  if (!hasNotes) return newNote();

  noteState.notes = fetchedNotes;

  sortStateNotes();

  const isSingleEmptyNote = fetchedNotes.length === 1 && isEmptyNote(fetchedNotes[0]);

  if (isSingleEmptyNote) {
    noteState.notes[0]!.timestamp = Date.now();
    // Clear these fields as whitespace-only can affect empty note checks
    noteState.notes[0]!.content.delta = {};
    noteState.notes[0]!.content.title = '';
    noteState.notes[0]!.content.body = '';
  }

  noteState.selectedNote = { ...noteState.notes[0]! };

  if (!isSingleEmptyNote) {
    clearEmptyNote();

    document.dispatchEvent(changeNoteEvent);
  }
}

/** Deletes note with the given `id`. */
export function deleteNote(id: string): void {
  noteState.notes.splice(findNoteIndex(id), 1);

  if (noteState.notes.length === 0) {
    newNote();
  } else if (noteState.selectedNote.id === id) {
    noteState.selectedNote = { ...noteState.notes[0]! };

    document.dispatchEvent(selectNoteEvent);
    document.dispatchEvent(changeNoteEvent);
  }

  if (syncState.unsyncedNoteIds.new === id) {
    syncState.unsyncedNoteIds.add({ new: '' });
  } else {
    document.dispatchEvent(getUnsyncedEvent(id, 'deleted'));

    tauriInvoke('delete_note', { id })
      .then(() => autoPush())
      .catch(console.error);
  }
}

/** Deletes {@link noteState.selectedNote} and all notes in {@link noteState.extraSelectedNotes}. */
export function deleteSelectedNotes(): void {
  deleteNote(noteState.selectedNote.id);

  noteState.extraSelectedNotes.forEach((nt) => {
    deleteNote(nt.id);
  });

  noteState.extraSelectedNotes = [];
}

/** Creates an empty note. */
export function newNote(isButtonClick?: boolean): void {
  const foundNote = findNote(noteState.selectedNote.id);

  // Only update timestamp if selected note is empty
  if (foundNote && isEmptyNote(foundNote)) {
    noteState.selectedNote.timestamp = Date.now();

    return;
  }

  const freshNote = new Note();

  noteState.notes.unshift(freshNote);
  noteState.selectedNote = { ...freshNote };

  // Ensure fresh note isn't overwritten on pull
  if (isButtonClick) {
    syncState.unsyncedNoteIds.add({ new: freshNote.id });
  }

  document.dispatchEvent(selectNoteEvent);
  document.dispatchEvent(changeNoteEvent);
  document.dispatchEvent(newNoteEvent);

  tauriInvoke('new_note', { note: { ...freshNote } }).catch(console.error);
}

/**
 * Edits currently selected note on Quill `text-change`.
 *
 * Selected note content should only be edited within `noteState.notes`, not
 * `noteState.selectedNote`, as `noteState.selectedNote` is what's reflected in the
 * editor. The timestamp, however, should be updated for both.
 */
export function editNote(delta: Partial<Delta>, title: string, body?: string): void {
  const foundNote = findNote(noteState.selectedNote.id);
  if (!foundNote || delta === foundNote.content.delta) return;

  const timestamp = Date.now();

  foundNote.timestamp = timestamp;
  noteState.selectedNote.timestamp = timestamp;

  foundNote.content = { delta, title, body: body || '' };

  sortStateNotes();

  document.dispatchEvent(getUnsyncedEvent(foundNote.id, 'edited'));

  tauriInvoke('edit_note', { note: { ...foundNote } })
    .then(autoPush)
    .catch(console.error);
}

/** Exports all notes, or a given selection. */
export async function exportNotes(noteIds: string[]): Promise<void> {
  const saveDir = await dialog.open({
    title: 'Choose a location',
    directory: true,
    multiple: false,
    recursive: false,
  });
  if (!saveDir) return;

  const notes = noteState.notes.filter((nt) => noteIds?.includes(nt.id));

  tauriInvoke('export_notes', { saveDir, notes }).catch(console.error);
}
