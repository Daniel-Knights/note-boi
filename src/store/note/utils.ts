import { Note } from '../../classes';
import { isEmptyNote } from '../../utils';

import { deleteNote } from './operations';
import { noteState } from './state';

/** Sorts notes in descending order by timestamp. */
export const sortNotesFn = (a: Note, b: Note): number => b.timestamp - a.timestamp;

/** Sorts {@link noteState.notes} in descending order by timestamp. */
export const sortStateNotes = (): Note[] => noteState.notes.sort(sortNotesFn);

/** Finds note index within {@link noteState.notes}. */
export function findNoteIndex(uuid?: string): number {
  return noteState.notes.findIndex((nt) => nt.uuid === uuid);
}

/** Finds note within {@link noteState.notes}. */
export function findNote(uuid?: string): Note | undefined {
  return noteState.notes.find((nt) => nt.uuid === uuid);
}

/**
 * Returns true if note is either {@link noteState.selectedNote}
 * or within {@link noteState.extraSelectedNotes}.
 */
export function isSelectedNote(note: Note): boolean {
  return (
    note.uuid === noteState.selectedNote.uuid ||
    noteState.extraSelectedNotes.some((nt) => nt?.uuid === note.uuid)
  );
}

/** Deletes {@link noteState.selectedNote} when note is empty. */
export function clearEmptyNote(): void {
  const isValidClear = noteState.notes.length > 1;
  if (!isValidClear) return;

  const foundNote = findNote(noteState.selectedNote.uuid);
  if (!foundNote) return;

  if (isEmptyNote(foundNote)) {
    deleteNote(noteState.selectedNote.uuid);
  }
}
