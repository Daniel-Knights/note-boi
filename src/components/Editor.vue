<template>
  <section id="editor">
    <small class="editor__date">{{
      unixToDateTime(state.selectedNote.timestamp || 0)
    }}</small>
    <div class="editor__body"></div>
  </section>
</template>

<script lang="ts" setup>
import { onMounted, ref, watchEffect } from 'vue';
import Quill from 'quill';

import { state, editBody } from '../store';
import { unixToDateTime, isEmptyNote } from '../utils';

const noteTitle = ref<HTMLElement | undefined>(undefined);

let quillEditor: Quill | undefined;
let isNoteSelect = false;

// Cleanup previously typed content
watchEffect(() => {
  if (!isEmptyNote(state.selectedNote)) return;

  if (noteTitle.value) noteTitle.value.innerText = '';
});

document.addEventListener('note-change', () => {
  const parsedBody = JSON.parse(state.selectedNote.body.delta || '[]');

  quillEditor?.setContents(parsedBody);
});

document.addEventListener('note-select', () => {
  isNoteSelect = true;
});

onMounted(() => {
  quillEditor = new Quill('.editor__body', {
    modules: {
      toolbar: true,
    },
    placeholder: 'Body',
    theme: 'snow',
  });

  quillEditor.on('text-change', () => {
    if (isNoteSelect) {
      isNoteSelect = false;
      return;
    }

    if (!quillEditor) return;
    const delta = quillEditor.getContents();
    const text = quillEditor.root.innerText;

    editBody(JSON.stringify(delta), text);
  });
});
</script>

<style lang="scss">
#editor {
  height: 100%;
  overflow-y: scroll;

  &::-webkit-scrollbar {
    display: none;
  }

  .editor__date,
  .ql-editor {
    padding-left: 8px;
    padding-right: 8px;
  }

  .editor__date {
    display: block;
    padding-top: 10px;
    padding-bottom: 10px;
    text-align: center;
    font-size: 11px;
    color: var(--color__tertiary);
    border-bottom: 1px solid var(--color__tertiary);
  }

  .ql-toolbar,
  .ql-container {
    border: none;
  }
}
</style>
