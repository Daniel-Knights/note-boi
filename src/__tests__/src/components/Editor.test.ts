import { mount } from '@vue/test-utils';

import * as n from '../../../store/note';
import { unixToDateTime } from '../../../utils';
import { mockApi } from '../../api';
import localNotes from '../../notes.json';
import { getByTestId } from '../../utils';

import Editor from '../../../components/Editor.vue';

describe('Editor', () => {
  it('Mounts', async () => {
    const { calls, events, promises } = mockApi();

    const wrapper = mount(Editor);

    await Promise.all(promises);

    assert.isTrue(wrapper.isVisible());
    assert.strictEqual(calls.length, 0);
    assert.strictEqual(events.emits.length, 0);
    assert.strictEqual(events.listeners.length, 0);
  });

  it('Has the correct timestamp', () => {
    const wrapper = mount(Editor);
    const timestamp = getByTestId(wrapper, 'timestamp');

    assert.strictEqual(timestamp.text(), unixToDateTime(new Date().getTime()));
  });

  it('Sets the correct note text', async () => {
    const { calls } = mockApi();
    const wrapper = mount(Editor);
    const editorBody = getByTestId(wrapper, 'body');

    assert.isEmpty(editorBody.text());

    await n.getAllNotes();

    assert.strictEqual(calls.length, 1);
    assert.isTrue(calls.has('get_all_notes'));
    assert.include(editorBody.text(), '¯\\_(ツ)_/¯');

    n.selectNote(localNotes[1].id);

    assert.include(editorBody.text(), localNotes[1].content.body);
  });
});
