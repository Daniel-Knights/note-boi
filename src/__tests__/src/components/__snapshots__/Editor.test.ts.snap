// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Editor > Has the correct timestamp > KeyStore.getKey() 1`] = `undefined`;

exports[`Editor > Has the correct timestamp > Storage.get('MENU_WIDTH') 1`] = `null`;

exports[`Editor > Has the correct timestamp > Storage.get('THEME') 1`] = `null`;

exports[`Editor > Has the correct timestamp > Storage.get('UPDATE_SEEN') 1`] = `null`;

exports[`Editor > Has the correct timestamp > Storage.get('UPDATE_STRATEGY') 1`] = `null`;

exports[`Editor > Has the correct timestamp > Storage.get('USERNAME') 1`] = `null`;

exports[`Editor > Has the correct timestamp > allCalls 1`] = `[]`;

exports[`Editor > Has the correct timestamp > document.body.outerHTML 1`] =
  `"<body class="theme--system"></body>"`;

exports[`Editor > Has the correct timestamp > mockDb 1`] = `
{
  "encryptedNotes": 10,
  "users": {
    "d": "1",
  },
}
`;

exports[`Editor > Has the correct timestamp > noteState 1`] = `
{
  "extraSelectedNotes": [],
  "notes": [],
  "selectedNote": {
    "content": {
      "body": "",
      "delta": {
        "ops": [],
      },
      "title": "",
    },
    "id": "id",
    "timestamp": 0,
  },
}
`;

exports[`Editor > Has the correct timestamp > openedPopup 1`] = `undefined`;

exports[`Editor > Has the correct timestamp > selectedTheme 1`] = `"System"`;

exports[`Editor > Has the correct timestamp > syncState 1`] = `
{
  "appError": AppError {
    "code": 0,
    "display": undefined,
    "message": undefined,
    "originalError": undefined,
    "retry": [Function],
    "retryConfig": undefined,
  },
  "isLoggedIn": false,
  "loadingCount": 0,
  "newPassword": "",
  "password": "",
  "unsyncedNoteIds": {
    "clear": [Function],
    "deleted": Set {},
    "edited": Set {},
    "new": "",
    "set": [Function],
    "size": 0,
    "store": [Function],
  },
  "username": "",
}
`;

exports[`Editor > Has the correct timestamp > updateState 1`] = `
{
  "isAvailable": false,
  "isDownloading": false,
  "strategy": "manual",
}
`;

exports[`Editor > Mounts > KeyStore.getKey() 1`] = `undefined`;

exports[`Editor > Mounts > Storage.get('MENU_WIDTH') 1`] = `null`;

exports[`Editor > Mounts > Storage.get('THEME') 1`] = `null`;

exports[`Editor > Mounts > Storage.get('UPDATE_SEEN') 1`] = `null`;

exports[`Editor > Mounts > Storage.get('UPDATE_STRATEGY') 1`] = `null`;

exports[`Editor > Mounts > Storage.get('USERNAME') 1`] = `null`;

exports[`Editor > Mounts > allCalls 1`] = `[]`;

exports[`Editor > Mounts > document.body.outerHTML 1`] =
  `"<body class="theme--system"></body>"`;

exports[`Editor > Mounts > mockDb 1`] = `
{
  "encryptedNotes": 10,
  "users": {
    "d": "1",
  },
}
`;

exports[`Editor > Mounts > noteState 1`] = `
{
  "extraSelectedNotes": [],
  "notes": [],
  "selectedNote": {
    "content": {
      "body": "",
      "delta": {
        "ops": [],
      },
      "title": "",
    },
    "id": "id",
    "timestamp": 0,
  },
}
`;

exports[`Editor > Mounts > openedPopup 1`] = `undefined`;

exports[`Editor > Mounts > selectedTheme 1`] = `"System"`;

exports[`Editor > Mounts > syncState 1`] = `
{
  "appError": AppError {
    "code": 0,
    "display": undefined,
    "message": undefined,
    "originalError": undefined,
    "retry": [Function],
    "retryConfig": undefined,
  },
  "isLoggedIn": false,
  "loadingCount": 0,
  "newPassword": "",
  "password": "",
  "unsyncedNoteIds": {
    "clear": [Function],
    "deleted": Set {},
    "edited": Set {},
    "new": "",
    "set": [Function],
    "size": 0,
    "store": [Function],
  },
  "username": "",
}
`;

exports[`Editor > Mounts > updateState 1`] = `
{
  "isAvailable": false,
  "isDownloading": false,
  "strategy": "manual",
}
`;

exports[`Editor > Sets the correct note text > KeyStore.getKey() 1`] = `undefined`;

exports[`Editor > Sets the correct note text > Storage.get('MENU_WIDTH') 1`] = `null`;

exports[`Editor > Sets the correct note text > Storage.get('THEME') 1`] = `null`;

exports[`Editor > Sets the correct note text > Storage.get('UPDATE_SEEN') 1`] = `null`;

exports[`Editor > Sets the correct note text > Storage.get('UPDATE_STRATEGY') 1`] =
  `null`;

exports[`Editor > Sets the correct note text > Storage.get('USERNAME') 1`] = `null`;

exports[`Editor > Sets the correct note text > allCalls 1`] = `
[
  [
    "invoke",
    {
      "calledWith": {},
      "name": "get_all_notes",
    },
  ],
]
`;

exports[`Editor > Sets the correct note text > document.body.outerHTML 1`] =
  `"<body class="theme--system"></body>"`;

exports[`Editor > Sets the correct note text > mockDb 1`] = `
{
  "encryptedNotes": 10,
  "users": {
    "d": "1",
  },
}
`;

exports[`Editor > Sets the correct note text > noteState 1`] = `
{
  "extraSelectedNotes": [],
  "notes": [
    {
      "content": {
        "body": "😬ö",
        "delta": {
          "ops": [
            {
              "insert": "Note with special characters

😬ö
  sss
	
Ketchup
Salt 
Sugar

¯\\_(ツ)_/¯

──────────────────────────────────
──────▄▀▀▄────────────────▄▀▀▄────
─────▐▒▒▒▒▌──────────────▌▒▒▒▒▌───
─────▌▒▒▒▒▐─────────────▐▒▒▒▒▒▐───
────▐▒▒▒▒▒▒▌─▄▄▄▀▀▀▀▄▄▄─▌▒▒▒▒▒▒▌──
───▄▌▒▒▒▒▒▒▒▀▒▒▒▒▒▒▒▒▒▒▀▒▒▒▒▒▒▐───
─▄▀▒▐▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▌───
▐▒▒▒▌▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▐───
▌▒▒▌▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▌──
▒▒▐▒▒▒▒▒▒▒▒▒▄▀▀▀▀▄▒▒▒▒▒▄▀▀▀▀▄▒▒▐──
▒▒▌▒▒▒▒▒▒▒▒▐▌─▄▄─▐▌▒▒▒▐▌─▄▄─▐▌▒▒▌─
▒▐▒▒▒▒▒▒▒▒▒▐▌▐█▄▌▐▌▒▒▒▐▌▐█▄▌▐▌▒▒▐─
▒▌▒▒▒▒▒▒▒▒▒▐▌─▀▀─▐▌▒▒▒▐▌─▀▀─▐▌▒▒▒▌
▒▌▒▒▒▒▒▒▒▒▒▒▀▄▄▄▄▀▒▒▒▒▒▀▄▄▄▄▀▒▒▒▒▐
▒▌▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▄▄▄▒▒▒▒▒▒▒▒▒▒▒▐
▒▌▒▒▒▒▒▒▒▒▒▒▒▒▀▒▀▒▒▒▀▒▒▒▀▒▀▒▒▒▒▒▒▐
▒▌▒▒▒▒▒▒▒▒▒▒▒▒▒▀▒▒▒▄▀▄▒▒▒▀▒▒▒▒▒▒▒▐
▒▐▒▒▒▒▒▒▒▒▒▒▀▄▒▒▒▄▀▒▒▒▀▄▒▒▒▄▀▒▒▒▒▐
▒▓▌▒▒▒▒▒▒▒▒▒▒▒▀▀▀▒▒▒▒▒▒▒▀▀▀▒▒▒▒▒▒▐
▒▓▓▌▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▌
▒▒▓▐▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▌─
▒▒▓▓▀▀▄▄▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▐──
▒▒▒▓▓▓▓▓▀▀▄▄▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▄▄▀▀▒▌─
▒▒▒▒▒▓▓▓▓▓▓▓▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▒▒▒▒▒▐─
▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▌
▒▒▒▒▒▒▒█▒█▒█▀▒█▀█▒█▒▒▒█▒█▒█▒▒▒▒▒▒▐
▒▒▒▒▒▒▒█▀█▒█▀▒█▄█▒▀█▒█▀▒▀▀█▒▒▒▒▒▒▐
▒▒▒▒▒▒▒▀▒▀▒▀▀▒▀▒▀▒▒▒▀▒▒▒▀▀▀▒▒▒▒▒▒▐
█▀▄▒█▀▄▒█▀▒█▀█▒▀█▀▒█▒█▒█▒█▄▒█▒▄▀▀▐
█▀▄▒█▀▄▒█▀▒█▄█▒▒█▒▒█▀█▒█▒█▀██▒█▒█▐
▀▀▒▒▀▒▀▒▀▀▒▀▒▀▒▒▀▒▒▀▒▀▒▀▒▀▒▒▀▒▒▀▀▐
▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▐

",
            },
          ],
        },
        "title": "Note with special characters",
      },
      "id": "id",
      "timestamp": 0,
    },
    {
      "content": {
        "body": "amet consectetur adipisicing",
        "delta": {
          "ops": [
            {
              "insert": "Amet consectetur adipisicing\\n\\namet consectetur adipisicing\\namet consectetur adipisicing\\namet consectetur adipisicing\\n",
            },
          ],
        },
        "title": "Amet consectetur adipisicing",
      },
      "id": "id",
      "timestamp": 0,
    },
    {
      "content": {
        "body": "amet consectetur adipisicing",
        "delta": {
          "ops": [
            {
              "insert": "Amet consectetur adipisicing\\n\\namet consectetur adipisicing\\namet consectetur adipisicing\\namet consectetur adipisicing\\n",
            },
          ],
        },
        "title": "Amet consectetur adipisicing",
      },
      "id": "id",
      "timestamp": 0,
    },
    {
      "content": {
        "body": "amet consectetur adipisicing",
        "delta": {
          "ops": [
            {
              "insert": "Amet consectetur adipisicing\\n\\namet consectetur adipisicing\\namet consectetur adipisicing\\namet consectetur adipisicing\\n",
            },
          ],
        },
        "title": "Amet consectetur adipisicing",
      },
      "id": "id",
      "timestamp": 0,
    },
    {
      "content": {
        "body": "amet consectetur adipisicing",
        "delta": {
          "ops": [
            {
              "insert": "Amet consectetur adipisicing\\n\\namet consectetur adipisicing\\namet consectetur adipisicing\\namet consectetur adipisicing\\n",
            },
          ],
        },
        "title": "Amet consectetur adipisicing",
      },
      "id": "id",
      "timestamp": 0,
    },
    {
      "content": {
        "body": "amet consectetur adipisicing",
        "delta": {
          "ops": [
            {
              "insert": "Amet consectetur adipisicing\\n\\namet consectetur adipisicing\\namet consectetur adipisicing\\namet consectetur adipisicing\\n",
            },
          ],
        },
        "title": "Amet consectetur adipisicing",
      },
      "id": "id",
      "timestamp": 0,
    },
    {
      "content": {
        "body": "amet consectetur adipisicing",
        "delta": {
          "ops": [
            {
              "insert": "Amet consectetur adipisicing\\n\\namet consectetur adipisicing\\namet consectetur adipisicing\\namet consectetur adipisicing\\n",
            },
          ],
        },
        "title": "Amet consectetur adipisicing",
      },
      "id": "id",
      "timestamp": 0,
    },
    {
      "content": {
        "body": "amet consectetur adipisicing",
        "delta": {
          "ops": [
            {
              "insert": "Amet consectetur adipisicing\\n\\namet consectetur adipisicing\\namet consectetur adipisicing\\namet consectetur adipisicing\\n",
            },
          ],
        },
        "title": "Amet consectetur adipisicing",
      },
      "id": "id",
      "timestamp": 0,
    },
    {
      "content": {
        "body": "amet consectetur adipisicing",
        "delta": {
          "ops": [
            {
              "insert": "Amet consectetur adipisicing\\n\\namet consectetur adipisicing\\namet consectetur adipisicing\\namet consectetur adipisicing\\n",
            },
          ],
        },
        "title": "Amet consectetur adipisicing",
      },
      "id": "id",
      "timestamp": 0,
    },
    {
      "content": {
        "body": "amet consectetur adipisicing",
        "delta": {
          "ops": [
            {
              "insert": "Amet consectetur adipisicing\\n\\namet consectetur adipisicing\\namet consectetur adipisicing\\namet consectetur adipisicing\\n",
            },
          ],
        },
        "title": "Amet consectetur adipisicing",
      },
      "id": "id",
      "timestamp": 0,
    },
  ],
  "selectedNote": {
    "content": {
      "body": "amet consectetur adipisicing",
      "delta": {
        "ops": [
          {
            "insert": "Amet consectetur adipisicing\\n\\namet consectetur adipisicing\\namet consectetur adipisicing\\namet consectetur adipisicing\\n",
          },
        ],
      },
      "title": "Amet consectetur adipisicing",
    },
    "id": "id",
    "timestamp": 0,
  },
}
`;

exports[`Editor > Sets the correct note text > openedPopup 1`] = `undefined`;

exports[`Editor > Sets the correct note text > selectedTheme 1`] = `"System"`;

exports[`Editor > Sets the correct note text > syncState 1`] = `
{
  "appError": AppError {
    "code": 0,
    "display": undefined,
    "message": undefined,
    "originalError": undefined,
    "retry": [Function],
    "retryConfig": undefined,
  },
  "isLoggedIn": false,
  "loadingCount": 0,
  "newPassword": "",
  "password": "",
  "unsyncedNoteIds": {
    "clear": [Function],
    "deleted": Set {},
    "edited": Set {},
    "new": "",
    "set": [Function],
    "size": 0,
    "store": [Function],
  },
  "username": "",
}
`;

exports[`Editor > Sets the correct note text > updateState 1`] = `
{
  "isAvailable": false,
  "isDownloading": false,
  "strategy": "manual",
}
`;
